custom/inddex/couch_db_data_collector.py

```diff
@@ -0,0 +1,133 @@
+            raise ValueError('Unknown \'table_name\'')
+        records = FixtureDataItem.by_data_type(self.domain, self.tables[table_name])
+
+        if not fields_and_values:
+            return records
+        values = {x[0]: x[1] for x in fields_and_values}
+
+        for record in records:
+            if self._is_record_valid(record, values):
+                if as_dict:
+                    data.append(self.record_to_dict(record))
+                else:
+                    data.append(record)
+
+        return data
+
+    @staticmethod
+    def record_to_dict(record):
+        """
+        transforms :param record from FixtureDataItem object into a dictionary
+        """
+        record_as_dict = {}
+        fields = record.fields
+        for field in fields:
+            field_list = fields[field].field_list
+            value = None if not field_list else field_list[0].field_value
+            record_as_dict[field] = value
+
+        return record_as_dict
+
+    @memoized
+    def get_data_from_table_as_dict(self, key_field, table_name=None, fields_and_values: Optional[Tuple] = None):
+        """
+        does the same filtering as get_data_from_table, but returns the data in
+        form of a dictionary with values of :param key_field as keys
+        """
+        if key_field not in (element[0] for element in fields_and_values):
+            raise ValueError('\'key_field\'s name must be in \'fields_and_values\'')
+        messy_dicts = self.get_data_from_table(
+            table_name=table_name, fields_and_values=fields_and_values, as_dict=True
+        )
+        clean_dicts = {}
+
+        for messy_dict in messy_dicts:
+            clean_dicts[messy_dict[key_field]] = self._get_dict(key_field, messy_dict)
+
+        return clean_dicts
+
+    @staticmethod
+    def _get_dict(id_key, dict_):
+        tmp_dict = dict_.copy()
+        tmp_dict.pop(id_key)
+
+        return tmp_dict
+
+    def _is_record_valid(self, record, values):
+        if values:
+            fields_values = self._get_fields_values_if_fields_exist(
+                record, [x for x in values.keys()]
+            )
+            if fields_values and self._values_match(fields_values, values):
+                return record
+
+        return None
+
+    @staticmethod
+    def _get_fields_values_if_fields_exist(record, fields):
+        values_for_field = {}
+        for field in fields:
+            field_ = record.fields.get(field)
+            if not field_:
+                return False
+            elif field_ and field_.field_list:
+                values_for_field[field] = field_.field_list[0].field_value
+
+        return values_for_field
+
+    @staticmethod
+    def _values_match(field_values, values):
+        matching_fields = set()
+        for key in values:
+            if field_values[key] in values[key]:
+                matching_fields.add(key)
+
+        return len(matching_fields) == len(field_values)
```

Comment:
If I understand correctly, `values` is a dictionary whose keys are field names, and whose values are going to be used to filter records. And `field_values` is a record converted to a dictionary, with only fields that are in the `values` dictionary. Is that right?

So we are checking that the value of each field is among the values we are selecting for? So if any of them don't match, we want to return False? If I'm understanding correctly, we can return False as soon as we don't get a match.

```suggestion
        for key in values:
            if field_values[key] not in values[key]:
                return False
        return True
```
... or better yet ...
```suggestion
        return all(field_values[key] in values[key] for key in values)
```